File: java_skeleton/javabot/Player.java
package javabot;

import javabot.skeleton.Action;
import javabot.skeleton.ActionType;
import javabot.skeleton.GameState;
import javabot.skeleton.State;
import javabot.skeleton.TerminalState;
import javabot.skeleton.RoundState;
import javabot.skeleton.Bot;
import javabot.skeleton.Runner;

import java.util.List;
import java.util.Random;
import java.util.Set;
import java.lang.Integer;
import java.lang.String;

/**
 * A pokerbot.
 */
public class Player implements Bot {
    // Your instance variables go here.

    /**
     * Called when a new game starts. Called exactly once.
     */
    public Player() {
    }

    /**
     * Called when a new round starts. Called State.NUM_ROUNDS times.
     *
     * @param gameState The GameState object.
     * @param roundState The RoundState object.
     * @param active Your player's index.
     */
    public void handleNewRound(GameState gameState, RoundState roundState, int active) {
        //int myBankroll = gameState.bankroll;  // the total number of chips you've gained or lost from the beginning of the game to the start of this round
        //float gameClock = gameState.gameClock;  // the total number of seconds your bot has left to play this game
        //int roundNum = gameState.roundNum;  // the round number from 1 to State.NUM_ROUNDS
        //List<String> myCards = roundState.hands.get(active);  // your cards
        //boolean bigBlind = (active == 1);  // true if you are the big blind
    }

    /**
     * Called when a round ends. Called State.NUM_ROUNDS times.
     *
     * @param gameState The GameState object.
     * @param terminalState The TerminalState object.
     * @param active Your player's index.
     */
    public void handleRoundOver(GameState gameState, TerminalState terminalState, int active) {
        //int myDelta = terminalState.deltas.get(active);  // your bankroll change from this round
        //RoundState previousState = (RoundState)(terminalState.previousState);  // RoundState before payoffs
        //int street = previousState.street;  // 0, 3, 4, or 5 representing when this round ended
        //List<String> myCards = previousState.hands.get(active);  // your cards
        //List<String> oppCards = previousState.hands.get(1-active);  // opponent's cards or "" if not revealed
    }

    /**
     * Where the magic happens - your code should implement this function.
     * Called any time the engine needs an action from your bot.
     *
     * @param gameState The GameState object.
     * @param roundState The RoundState object.
     * @param active Your player's index.
     * @return Your action.
     */
    public Action getAction(GameState gameState, RoundState roundState, int active) {

        // May be helpful, but you can choose to not use.
        Set<ActionType> legalActions = roundState.legalActions();  // the actions you are allowed to take
        int street = roundState.street;  // 0, 3, 4, or 5 representing pre-flop, flop, turn, or river respectively
        List<String> myCards = roundState.hands.get(active);  // your cards
        List<String> boardCards = roundState.deck;  // the board cards
        int myPip = roundState.pips.get(active);  // the number of chips you have contributed to the pot this round of betting
        int oppPip = roundState.pips.get(1-active);  // the number of chips your opponent has contributed to the pot this round of betting
        int myStack = roundState.stacks.get(active);  // the number of chips you have remaining
        int oppStack = roundState.stacks.get(1-active);  // the number of chips your opponent has remaining
        Integer myBid = roundState.bids.get(active); // How much you have bid. Null or -1 until after the auction.
        Integer oppBid = roundState.bids.get(1-active); // How much your opponent has bid. Null or -1 until after auction.
        int continueCost = oppPip - myPip;  // the number of chips needed to stay in the pot
        int myContribution = State.STARTING_STACK - myStack;  // the number of chips you have contributed to the pot
        int oppContribution = State.STARTING_STACK - oppStack;  // the number of chips your opponent has contributed to the pot
        int minCost = 0;
        int maxCost = 0;
        if (legalActions.contains(ActionType.RAISE_ACTION_TYPE)) {
            List<Integer> raiseBounds = roundState.raiseBounds(); // the smallest and largest numbers of chips for a legal bet/raise
            minCost = raiseBounds.get(0) - myPip; // the cost of a minimum bet/raise
            maxCost = raiseBounds.get(1) - myPip; // the cost of a maximum bet/raise
        }

        // Basic bot that bids and raises randomly, or just checks and calls.
        Random rand = new Random();
        if (legalActions.contains(ActionType.BID_ACTION_TYPE)) { 
            return new Action(ActionType.BID_ACTION_TYPE, rand.nextInt(myStack)); // Random bid between 0 and our stack size
        }
        if (legalActions.contains(ActionType.CHECK_ACTION_TYPE)) {  // Check
            return new Action(ActionType.CHECK_ACTION_TYPE);
        } else {
            return new Action(ActionType.CALL_ACTION_TYPE);
        }



    }

    /**
     * Main program for running a Java pokerbot.
     */
    public static void main(String[] args) {
        Player player = new Player();
        Runner runner = new Runner();
        runner.parseArgs(args);
        runner.runBot(player);
    }
}

File: python_skeleton/player.py
'''
Simple example pokerbot, written in Python.
'''
from skeleton.actions import FoldAction, CallAction, CheckAction, RaiseAction, BidAction
from skeleton.states import GameState, TerminalState, RoundState
from skeleton.states import NUM_ROUNDS, STARTING_STACK, BIG_BLIND, SMALL_BLIND
from skeleton.bot import Bot
from skeleton.runner import parse_args, run_bot
import random
import eval7


class Player(Bot):
    '''
    A pokerbot.
    '''

    def __init__(self):
        '''
        Called when a new game starts. Called exactly once.

        Arguments:
        Nothing.

        Returns:
        Nothing.
        '''
        pass

    def handle_new_round(self, game_state, round_state, active):
        '''
        Called when a new round starts. Called NUM_ROUNDS times.

        Arguments:
        game_state: the GameState object.
        round_state: the RoundState object.
        active: your player's index.

        Returns:
        Nothing.
        '''
        my_bankroll = game_state.bankroll  # the total number of chips you've gained or lost from the beginning of the game to the start of this round
        game_clock = game_state.game_clock  # the total number of seconds your bot has left to play this game
        round_num = game_state.round_num  # the round number from 1 to NUM_ROUNDS
        my_cards = round_state.hands[active]  # your cards
        big_blind = bool(active)  # True if you are the big blind


        monte_carlo_iters = 100
        game_clock = game_state.game_clock
        round_num = game_state.round_num

        self.strength = self.calculate_strength(my_cards, monte_carlo_iters)

        #cnan use this to see number of iterations we can squeeze in
        if round_num == NUM_ROUNDS:
            print(game_clock)


        pass

    def calculate_strength(self, hole, iters):
        deck = eval7.Deck()
        hole_cards = [eval7.Card(card) for card in hole]
        for card in hole_cards:
            deck.cards.remove(card)

        score_with_card = 0
        score_without_card = 0

        
        for _ in range(iters):
            deck.shuffle()

            com = 5
            opp = 3
            draw = deck.peek(com + opp)

            opp_hole = draw[:opp]
            community_cards = draw[opp:]
            our_hand = hole_cards + community_cards
            opp_hand = opp_hole + community_cards
            our_hand_value = eval7.evaluate(our_hand)
            opp_hand_value = eval7.evaluate(opp_hand)

            if our_hand_value > opp_hand_value:
                    #we win
                score_without_card += 2
            elif our_hand_value == opp_hand_value:
                #we tie
                score_without_card += 1
            else:
                #we lost
                score_without_card += 0

        for _ in range(iters):
            deck.shuffle()

            ours = 1
            com = 5
            opp = 2
            draw = deck.peek(com + opp + ours)

            opp_hole = draw[:opp]
            community_cards = draw[opp:opp+com]
            our_draw = draw[opp+com:]

            our_hand = hole_cards + our_draw + community_cards
            opp_hand = opp_hole + community_cards
            our_hand_value = eval7.evaluate(our_hand)
            opp_hand_value = eval7.evaluate(opp_hand)

            if our_hand_value > opp_hand_value:
                #we win
                score_with_card += 2
            elif our_hand_value == opp_hand_value:
                #we tie
                score_with_card += 1
            else:
                #we lost
                score_with_card += 0

        strength_without_card = score_without_card/(2*iters)
        strength_with_card = score_with_card/(2*iters)

        avg_strength = (strength_with_card + strength_without_card)/2

        print(avg_strength)
        return avg_strength



    def handle_round_over(self, game_state, terminal_state, active):
        '''
        Called when a round ends. Called NUM_ROUNDS times.

        Arguments:
        game_state: the GameState object.
        terminal_state: the TerminalState object.
        active: your player's index.

        Returns:
        Nothing.
        '''

        #my_delta = terminal_state.deltas[active]  # your bankroll change from this round
        #previous_state = terminal_state.previous_state  # RoundState before payoffs
        #street = previous_state.street  # 0, 3, 4, or 5 representing when this round ended
        #my_cards = previous_state.hands[active]  # your cards
        #opp_cards = previous_state.hands[1-active]  # opponent's cards or [] if not revealed
        pass

    def get_action(self, game_state, round_state, active):
        '''
        Where the magic happens - your code should implement this function.
        Called any time the engine needs an action from your bot.

        Arguments:
        game_state: the GameState object.
        round_state: the RoundState object.
        active: your player's index.

        Returns:
        Your action.
        '''
        # May be useful, but you may choose to not use.
        legal_actions = round_state.legal_actions()  # the actions you are allowed to take
        street = round_state.street  # 0, 3, 4, or 5 representing pre-flop, flop, turn, or river respectively
        my_cards = round_state.hands[active]  # your cards
        board_cards = round_state.deck[:street]  # the board cards
        my_pip = round_state.pips[active]  # the number of chips you have contributed to the pot this round of betting
        opp_pip = round_state.pips[1-active]  # the number of chips your opponent has contributed to the pot this round of betting
        my_stack = round_state.stacks[active]  # the number of chips you have remaining
        opp_stack = round_state.stacks[1-active]  # the number of chips your opponent has remaining
        my_bid = round_state.bids[active]  # How much you bid previously (available only after auction)
        opp_bid = round_state.bids[1-active]  # How much opponent bid previously (available only after auction)
        continue_cost = opp_pip - my_pip  # the number of chips needed to stay in the pot
        my_contribution = STARTING_STACK - my_stack  # the number of chips you have contributed to the pot
        opp_contribution = STARTING_STACK - opp_stack  # the number of chips your opponent has contributed to the pot
        pot = my_contribution + opp_contribution
       
        strength = self.strength

        
        min_raise, max_raise = round_state.raise_bounds()  # the smallest and largest numbers of chips for a legal bet/raise

        if street < 3:
            raise_amt = int(my_pip + continue_cost + (0.4)*(pot+continue_cost))
        else:
            raise_amt = int(my_pip + continue_cost + (0.4)*(pot+continue_cost))

        raise_amt = max(min_raise,raise_amt)
        raise_amt = min(max_raise, raise_amt)
        raise_cost = raise_amt - my_pip

        if RaiseAction in legal_actions and raise_cost <= my_stack:
            commit_action = RaiseAction(raise_amt)
                        
        elif CallAction in legal_actions:
            commit_action = CallAction()
        else:
            commit_action = CheckAction()

        if continue_cost > 0:
            pot_odds = continue_cost/(pot + continue_cost)

            if strength >= pot_odds:
                if strength > 0.5 and random.random() < strength:
                    my_action = commit_action
                else:
                    my_action = CallAction()
                
            else:
                if random.random() < strength:
                        my_action = commit_action
                else:
                     my_action = CheckAction()
        else: # board continue cost == 0, so its our action
            if random.random() < strength:
                my_action = commit_action
            else:
                my_action = CheckAction()

        return my_action


        # if RaiseAction in legal_actions:
        #    min_raise, max_raise = round_state.raise_bounds()  # the smallest and largest numbers of chips for a legal bet/raise
        #    min_cost = min_raise - my_pip  # the cost of a minimum bet/raise
        #    max_cost = max_raise - my_pip  # the cost of a maximum bet/raise
        
        # if RaiseAction in legal_actions and random.random() < 0.3:
        #     return RaiseAction(random.randint(min_raise, max_raise))
        # if CheckAction in legal_actions:
        #     return CheckAction()
        # elif BidAction in legal_actions:
        #     return BidAction(my_stack) # random bid between 0 and our stack
        # return CallAction()


if __name__ == '__main__':
    run_bot(Player(), parse_args())
